### PR TITLE
Gif limiter

### PR DESCRIPTION
**Pull request by rm-rf-tux**
## Description
This PR intends to add a configurable GIF ratelimit to prevent the excessive use of these often low quality posts. 
It has two mechanisms of action:
- A channel limit: if too many GIFs were recently sent in a channel recently, new ones get deleted
- A user limit. Each channel can also have a configured per-user limit. If a user sends a GIF (anywhere except for in a blacklisted channel), it will be added to their GIF count. If they try sending another GIF, their count will be compared to the per-user maximum in that channel, and the GIF deleted if said quota is reached. This behavior prevents people moving to another channel to continue spamming GIFs.

## Guidelines

- My code follows the style guidelines of this project (formatted with Ruff)
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation if needed
- My changes generate no new warnings
- I have tested this change
- Any dependent changes have been merged and published in downstream modules
- I have added all appropriate labels to this PR

- [X] I have followed all of these guidelines.

## How Has This Been Tested? (if applicable)
Sending SPAM GIFs in the tux dev server and verifying that all ratelimits match what is configured

### CODE REVIEW

**Comment by amyulating** [comment]
now enough variables are called "meow" or "nya". fix by EOD or will not merge.

**Comment by rm-rf-tux** [comment]
> now enough variables are called "meow" or "nya". fix by EOD or will not merge.

Sure, but first make a PR https://github.com/torvalds/linux <here with the proposed changes. If it is pulled, I'll apply the changes to this branch.

File: config/settings.yml.example
Comment (electron271) [inline config/settings.yml.example]: any reason why these were changed

File: config/settings.yml.example
Comment (electron271) [inline config/settings.yml.example]: make a section for the gif limiter like for the user ids

File: tux/cogs/services/gif_limiter.py
Comment (electron271) [inline tux/cogs/services/gif_limiter.py]: for consistency put the docstring in the GifLimiter class

File: tux/cogs/services/gif_limiter.py
Comment (electron271) [reply tux/cogs/services/gif_limiter.py:88]: agreed

File: tux/utils/constants.py
Comment (electron271) [reply tux/utils/constants.py]: agreed

**Review by electron271 — CHANGES_REQUESTED** [review]

File: config/settings.yml.example
Comment (rm-rf-omega) [reply config/settings.yml.example]: Accidentally moved instead of copied the .example file
Before committing I recreated a .example file.

Comment (rm-rf-omega) [reply config/settings.yml.example]: Also this is an old account I found in my accounts list, sorry for the inconsistency

File: tux/cogs/services/gif_limiter.py
Comment (rm-rf-tux) [reply tux/cogs/services/gif_limiter.py:88]: > agreed

I'll do it but don't understand the point, we're talking about three early exit conditions in one if statement.

**Review by electron271 — APPROVED** [review]
looks good to me, though id appreciate a second review before merging

**Review by ghost — APPROVED** [review]
Why do we need `asyncio.Lock()` in the cleanup task? Seems overly aggressive for a GIF remove check. You could do something like this to emit the 20-second loop entirely. Not sure how much this will affect efficiency. 

```py
    async def _handle_gif_message(self, message: discord.Message) -> None:
        ...
        # Remove old GIF timestamps
        self._remove_old_gifs(channel, current_time)
        self._remove_old_gifs(user, current_time, is_user=True)
		...

    def _remove_old_gifs(self, key: int, current_time: int, is_user: bool = False) -> None:
        """Removes old GIF timestamps for a given key (channel or user)"""
        gif_dict = self.recent_gifs_by_user if is_user else self.recent_gifs_by_channel
        if key in gif_dict:
            gif_dict[key] = [t for t in gif_dict[key] if current_time - t < self.recent_gif_age]
            if not gif_dict[key]:
                del gif_dict[key]
```

The code looks perfect otherwise so ready for merge if this isn't a solution you're looking for. 



</details>

**Comment by rm-rf-tux** [comment]
> Why do we need `asyncio.Lock()` in the cleanup task? Seems overly aggressive for a GIF remove check. You could do something like this to emit the 20-second loop entirely. Not sure how much this will affect efficiency.

And so I originally thought, because this is async code, not parallel execution. However THIS somehow happened: 
https://discord.com/channels/1184229800332771328/1223687134830133388/1288797709037211672
Clearly the on_message function executed while the cleanup code was running. It caused the task to fail which would effectively block GIFs entirely. Not a good idea.

Also parsing the entire GIF timestamp list at every flagged message would be expensive, and not very scalable if there are hundreds of active users. The lock is probably more CPU-friendly

**Comment by wlinator** [comment]
I agree, thanks for the clarification. I'll let @kzndotsh merge, in case she has any input before we add this.

**Comment by rm-rf-tux** [comment]
I suspect that an empty channel exclude list causes this. Shouldn't be a huge issue but will be fixed

**Pull request merged**
